app/models/appointment.rb

```diff
@@ -29,25 +32,33 @@ class Appointment < ApplicationRecord
   scope :waiting, -> { not_checked_in.not_checked_out }
 
   scope :available_doses, lambda {
-                            active_ubs
-                              .active
-                              .waiting
-                              .not_scheduled
-                          }
+    active_ubs
+      .active
+      .waiting
+      .not_scheduled
+  }
 
   scope :search_for, lambda { |text|
     joins(:patient)
       .where(
         Patient.arel_table[:cpf]
-          .eq(Patient.parse_cpf(text)) # Search for CPF without . and -
-          .or(Patient.arel_table[:name].matches("%#{text.strip}%"))
+               .eq(Patient.parse_cpf(text)) # Search for CPF without . and -
+               .or(Patient.arel_table[:name].matches("%#{text.strip}%"))
       )
   }
 
-  # In the case it's a second dose, patient shouldn't be able to reschedule the appointment.
-  # NOTE: in the future once we have a better way to handle 2nd doses, check against the recommended vaccination window
+  # For follow ups, can only be canceled or rescheduled close to the date.
   def can_cancel_and_reschedule?
-    patient.doses.empty? || start < Rails.configuration.x.schedule_up_to_days.days.from_now.end_of_day
+    return true unless follow_up_for_dose
+
+    Time.zone.now > can_change_after
+  end
+
+  # Patients can only cancel or reschedule after a certain cutoff, in this case being "schedule_up_to_days" related with
+  # when they should get the vaccine.
+  def can_change_after
+    follow_up_for_dose.created_at + follow_up_for_dose.follow_up_in_days.days -
+      Rails.configuration.x.schedule_up_to_days.days
```

Comment:
Acho que precisamos validar esta data, pois dessa forma está atrelada com o número de dias adiante que se é possível se agendar, que neste momento é de 10 dias, ou seja, caso o paciente tome a primeira dose da CoronaVac, ele poderá reagendar sua segunda dose a partir do dia 20 já.

É só uma questão de validar com o Denis / Secretaria. A alternativa seria colocar um valor fixo de dias, por exemplo: `follow_up_for_dose.created_at + follow_up_for_dose.follow_up_in_days.days - 5`

Reply:
É pelo que entendo o escopo de poder alterar é diferente do escopo 'estar próximo do relógio biológico' ou algo do tipo haha. Talvez ter uma env pra isso. Mas acho que essa do relógio biológico vai ter que sempre ser menor do que esta env do 'poder alterar' que é mais do ritmo de vacinação.

Reply:
A idéia é a pessoa poder alterar a data a partir de quando a data da D2 for acessível por reagendamento. Concordo que a Secretaria poderia dar uma indicação disso, eu já comentei algumas vezes lá no canal sobre esta data mas não tive resposta.